### ISSUE TITLE
`array.new_canon` with i8 element type is accepted by reference interpreter

### ISSUE THREAD

**Issue by catamorphism** [WebAssembly/gc#318]
The following code (provided by @takikawa ) runs in the reference interpreter and returns 0:

```
(module
  (type $a (array (mut i8)))
  (func (export "f") (result i32)
    (array.get_u $a
                 (array.new_canon $a (i32.const 256) (i32.const 4))
                 (i32.const 0))))

(invoke "f")
```

However, according to the spec, this should not be type-correct. From https://github.com/WebAssembly/gc/blob/main/proposals/gc/MVP.md -

```
array.new_canon $t : [t' i32] -> [(ref $t)]
iff expand($t) = array (mut t')
```

In the example code, `expand($t)` is `array (mut i8)`, so `t'` is `i8`. However, `i8` doesn't match the type of `(i32.const 256)`, which is `i32`.

This could be fixed in the spec as follows:

```
array.new_canon $t : [t'' i32] -> [(ref $t)]
iff expand($t) = array (mut t')
and unpacked(t') = t''
```

but I'm not sure if that was the intent.

**Comment by titzer** [comment]
I think it is intended to typecheck, because `array.set` on a packed array will also accept `i32` values, just implicitly narrow them.

**Comment by rossberg** [comment]
Yes, thanks for spotting, this was an oversight in the doc. Fixed by #319, for this and a couple of other instructions.
